internal/services/topic.go

```diff
@@ -312,14 +314,55 @@ func (ts *TopicService) requestedAssignments(currentPartitions int, brokers []*s
 		return brokers[i].ID() < brokers[j].ID()
 	})
 
-	assignments := make(map[int32][]int32, int(partitions))
+	rackMap := make(map[string][]*sarama.Broker)
+	var rackNames []string
+	brokersWithRack := 0
+	for _, broker := range brokers {
+		if broker.Rack() != "" {
+			brokersWithRack++
+			if _, ok := rackMap[broker.Rack()]; !ok {
+				rackMap[broker.Rack()] = make([]*sarama.Broker, 0)
+				rackNames = append(rackNames, broker.Rack())
+			}
+			rackMap[broker.Rack()] = append(rackMap[broker.Rack()], broker)
+		}
+	}
+
+	if len(brokers) != brokersWithRack {
+		if brokersWithRack > 0 {
+			glog.Warningf("Not *all* brokers have rack assignments (%d/%d), topic %s will be created without rack awareness", brokersWithRack, len(brokers), ts.canaryConfig.Topic)
+		}
```

Comment:
I was wondering if we could move the ordering of brokers IDs here instead of having that operation outside.
I mean, ordering by ID is not useful if we are going to use rack-awareness but only if we don't use rack-awareness. Wdyt?

Reply:
You are referring to `sort.Slice`, right?  This is still needed even in the rack case as it will ensure that ordering of brokers within the rack is defined too.  Without this the algorithm won't necessarily generate stable results.

Reply:
You are right!